### PR TITLE
Mve3 topic

### PR DESCRIPTION
**Pull request by romann-broque**

### CODE REVIEW

File: srcs/Channel.cpp [hunk 1]
```diff
@@ -23,7 +23,7 @@ static std::string getSpecifiedNick(const SpecifiedClient &spClient) {
 // Methods
 
 Channel::Channel(const std::string &name, const Client *const client)
-	: _name(name) {
+	: _name(name), _isTopicProtected(false) {
```
Comment (HelloThere-ImMat) [inline srcs/Channel.cpp]: mettre a true et changer les tests

File: srcs/Channel.cpp [hunk 2]
```diff
@@ -75,12 +77,37 @@ void Channel::sendToAll(
...
+			 userMap.begin();
 		 it != userMap.end(); it++) {
 		if (it->second.client->getNickname() == nick)
 			return (true);
 	}
 	return (false);
 }
+
+bool Channel::canChangeTopic(const Client *const client) const {
+	return (_isTopicProtected == true || isOp(client));
```
Comment (HelloThere-ImMat) [inline srcs/Channel.cpp]: mettre a false

File: srcs/Channel.cpp [hunk 3]
```diff
@@ -54,7 +54,9 @@ void Channel::removeUser(const Client *const client) {
 	userMap.erase(nickname);
 }
 
-const std::string &Channel::getName() const { return _name; }
+const std::string &Channel::getTopic() const { return _topic; }
```
Comment (HelloThere-ImMat) [inline srcs/Channel.cpp:57]: is it necessary, check and remove if not

File: srcs/Server.cpp [hunk 4]
```diff
@@ -42,10 +42,14 @@ Server::Server(const std::string &port, const std::string &password)
 	_cmdMap["JOIN"] = &Server::join;
 	_cmdMap["PRIVMSG"] = &Server::privmsg;
 	_cmdMap["PART"] = &Server::part;
+	_cmdMap["TOPIC"] = &Server::topic;
 
 	printLog("Port: " + port);
 	printLog("Password: " + password);
 	_epollFd = 0;
```
Comment (HelloThere-ImMat) [inline srcs/Server.cpp]: to be removed

File: srcs/ServerCmds.cpp [hunk 5]
```diff
@@ -34,13 +34,14 @@ static bool isNicknameValid(const std::string &nickname) {
...
 	std::string	 fullMessage;
 	const size_t size = cmd.size();
-	size_t		 i = PRIVMSG_START_INDEX;
-	if (size > 2) {
+	size_t		 i = startIndex;
+	if (size > startIndex) {
 		while (i < size) {
-			if (i > PRIVMSG_START_INDEX)
+			if (i > startIndex)
 				fullMessage += " ";
```
Comment (HelloThere-ImMat) [inline srcs/ServerCmds.cpp]: mettre un define

File: srcs/ServerCmds.cpp [hunk 6]
```diff
@@ -138,29 +138,68 @@ void Server::join(const std::vector<std::string> &cmd, Client *const client) {
...
-		if (channel->userIsInChannel(client))
-			channel->sendToOthers(
-				client, PRIVMSG_PREFIX + cmd[1] + " " + fullMessage);
+	const std::string fullMessage = getFullMessage(cmd, PRIVMSG_START_INDEX);
+	const std::string privMessage = PRIVMSG_PREFIX + cmd[1] + " " + fullMessage;
+	const std::map<std::string, Channel *>::iterator it =
+		_channels.find(cmd[1]);
+
+	if (it != _channels.end()) {
+		Channel *channel = it->second;
```
Comment (HelloThere-ImMat) [inline srcs/ServerCmds.cpp]: const ptr

File: srcs/ServerCmds.cpp [hunk 7]
```diff
@@ -138,29 +138,68 @@ void Server::join(const std::vector<std::string> &cmd, Client *const client) {
...
+		SendCmd::sendPrivateMessage(
+			privMessage, client, _clientMap.getClient(cmd[1]));
 	else
 		printLog("Cound not send message");
 }
 
 void Server::part(const std::vector<std::string> &cmd, Client *const client) {
-	if (_channels.find(CHANNEL_PREFIX + cmd[1]) != _channels.end()) {
-		Channel *channel = _channels.find(CHANNEL_PREFIX + cmd[1])->second;
+	if (_channels.find(cmd[1]) != _channels.end()) {
```
Comment (HelloThere-ImMat) [inline srcs/ServerCmds.cpp]: can be defined in an it above

File: srcs/ServerCmds.cpp [hunk 8]
```diff
@@ -138,29 +138,68 @@ void Server::join(const std::vector<std::string> &cmd, Client *const client) {
...
+		Channel *channel = _channels.find(cmd[1])->second;
 		channel->sendToAll(client, PART_PREFIX + cmd[1]);
 		channel->removeUser(client);
 	} else
 		printLog("Could not part channel");
 }
 
+static std::string getTopic(const std::string &message) {
```
Comment (HelloThere-ImMat) [inline srcs/ServerCmds.cpp]: rendre plus generique, removeSetter

File: srcs/Channel.cpp (hunk 3 above)
Comment (romann-broque) [reply srcs/Channel.cpp:57]: yes it is necessary

**Review by HelloThere-ImMat — APPROVED** [review]

**Pull request merged**